gadopt/transport_solver.py

```diff
@@ -0,0 +1,368 @@
+        List of equation terms (refer to terms_mapping)
+      solution:
+        Firedrake function for the field of interest
+      delta_t:
+        Simulation time step
+      timestepper:
+        Runge-Kutta time integrator employing an explicit or implicit numerical scheme
+      solution_old:
+        Firedrake function holding the solution field at the previous time step
+      eq_attrs:
+        Dictionary of terms arguments and their values
+      bcs:
+        Dictionary specifying boundary conditions (identifier, type, and value)
+      solver_parameters:
+        Dictionary of solver parameters or a string specifying a default configuration
+        provided to PETSc
+      su_diffusivity:
+        Float activating the streamline-upwind stabilisation scheme and specifying the
+        corresponding diffusivity
+
+    """
+
+    def __init__(
+        self,
+        terms: str | list[str],
+        solution: Function,
+        /,
+        delta_t: Constant,
+        timestepper: RungeKuttaTimeIntegrator,
+        **kwargs,
+    ) -> None:
+        self.terms = [terms] if isinstance(terms, str) else terms
+
+        super().__init__(solution, delta_t, timestepper, **kwargs)
+
+    def set_equation(self) -> None:
+        self.equation = Equation(
+            self.test,
+            self.solution_space,
+            residual_terms=[self.terms_mapping[term] for term in self.terms],
+            mass_term=scalar_eq.mass_term,
+            eq_attrs=self.eq_attrs,
+            bcs=self.weak_bcs,
+        )
+
+
+class EnergySolver(GenericTransportBase):
+    """Advances in time the energy conservation equation.
+
+    Note: The solution field is updated in place.
+
+    Args:
+      solution:
+        Firedrake function for temperature
+      u:
+        Firedrake function for velocity
+      approximation:
+        G-ADOPT approximation defining terms in the system of equations
+      delta_t:
+        Simulation time step
+      timestepper:
+        Runge-Kutta time integrator employing an explicit or implicit numerical scheme
+      solution_old:
+        Firedrake function holding the solution field at the previous time step
+      bcs:
+        Dictionary specifying boundary conditions (identifier, type, and value)
+      solver_parameters:
+        Dictionary of solver parameters or a string specifying a default configuration
+        provided to PETSc
+      su_diffusivity:
+        Float activating the streamline-upwind stabilisation scheme and specifying the
+        corresponding diffusivity
+
+    """
+
+    def __init__(
+        self,
+        solution: Function,
+        u: Function,
+        approximation: BaseApproximation,
+        /,
+        delta_t: Constant,
+        timestepper: RungeKuttaTimeIntegrator,
+        **kwargs,
+    ) -> None:
+        self.u = u
+        self.approximation = approximation
+
+        super().__init__(solution, delta_t, timestepper, **kwargs)
+
+        self.T_old = self.solution_old
```

Comment:
This is how `__init__` in `EnergySolver` looks like without the metaclass. I am not a fan of it, but do other devs prefer it better like that?